1110-i386-txt-Add-Intel-TXT-core-implementation.patch

```diff
@@ -0,0 +1,1038 @@
++
++  /* Choose monitor RLP wakeup mechanism first. */
++  if (sinit_caps & GRUB_TXT_CAPS_MONITOR_SUPPORT)
++    os_sinit_data->capabilities |= GRUB_TXT_CAPS_MONITOR_SUPPORT;
++  else if (sinit_caps & GRUB_TXT_CAPS_GETSEC_WAKE_SUPPORT)
++    os_sinit_data->capabilities |= GRUB_TXT_CAPS_GETSEC_WAKE_SUPPORT;
++  else
++    return grub_error (GRUB_ERR_BAD_ARGUMENT, N_("lack of RLP wakeup mechanism"));
++
++  if (sinit_caps & GRUB_TXT_CAPS_ECX_PT_SUPPORT)
++    os_sinit_data->capabilities |= GRUB_TXT_CAPS_ECX_PT_SUPPORT;
++
++  if (grub_get_tpm_ver () == GRUB_TPM_12)
++    return grub_error (GRUB_ERR_NOT_IMPLEMENTED_YET,
++		       N_("TPM 1.2 detected, but not implemented yet"));
++  else
++    {
++      if (!(sinit_caps & GRUB_TXT_CAPS_TPM_20_EVTLOG_SUPPORT))
++	return grub_error (GRUB_ERR_BAD_ARGUMENT,
++			   N_("original TXT TPM 2.0 event log format is not supported"));
++
++      os_sinit_data->capabilities |= GRUB_TXT_CAPS_TPM_20_EVTLOG_SUPPORT;
++
++      os_sinit_data->flags = GRUB_TXT_PCR_EXT_MAX_PERF_POLICY;
++
++      os_sinit_data->version = OS_SINIT_DATA_TPM_20_VER;
++
++      heap_event_log_pointer2_1_element =
++	(struct grub_txt_heap_event_log_pointer2_1_element *) os_sinit_data->ext_data_elts;
++      heap_event_log_pointer2_1_element->type = GRUB_TXT_HEAP_EXTDATA_TYPE_EVENT_LOG_POINTER2_1;
++      heap_event_log_pointer2_1_element->size = sizeof (*heap_event_log_pointer2_1_element);
++      heap_event_log_pointer2_1_element->phys_addr = slparams->tpm_evt_log_base;
++      heap_event_log_pointer2_1_element->allocated_event_container_size = slparams->tpm_evt_log_size;
++
++      heap_end_element = (struct grub_txt_heap_end_element *)
++	((grub_addr_t) heap_event_log_pointer2_1_element + heap_event_log_pointer2_1_element->size);
++      heap_end_element->type = GRUB_TXT_HEAP_EXTDATA_TYPE_END;
++      heap_end_element->size = sizeof (*heap_end_element);
++    }
++  grub_dprintf ("slaunch", "TXT HEAP init done\n");
++  /*
++   * TODO: TXT spec: Note: BiosDataSize + OsMleDataSize + OsSinitDataSize + SinitMleDataSize
++   * must be less than or equal to TXT.HEAP.SIZE, TXT spec, p. 102.
++   */
++
++  return GRUB_ERR_NONE;
++}
++
++/*
++ * TODO: Why 1 GiB limit? It does not seem that it is required by TXT spec.
++ * If there is a limit then it should be checked before allocation and image load.
++ *
++ * If enough room is available in front of the MLE, the maximum size of an
++ * MLE that can be covered is 1G. This is due to having 512 PDEs pointing
++ * to 512 page tables with 512 PTEs each.
++ */
++grub_uint32_t
++grub_txt_get_mle_ptab_size (grub_uint32_t mle_size)
++{
++  /*
++   * #PT + 1 PT + #PD + 1 PD + 1 PDT
++   *
++   * Why do we need 2 extra PTEs and PDEs? Yes, because MLE image may not
++   * start and end at PTE (page) and PDE (2 MiB) boundary...
++   */
++  return ((((mle_size / GRUB_PAGE_SIZE) + 2) / 512) + 1 +
++	  (((mle_size / (512 * GRUB_PAGE_SIZE)) + 2) / 512) + 1 + 1) * GRUB_PAGE_SIZE;
```

Comment:
Please break this up into multiple statements with intermediate variables so that this is easier to read.